transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,806 @@
+                    .handle(Instant::now(), peer, None, self.buffer[..bytes].into())
+                {
+                    Some(e) => e,
+                    None => continue,
+                };
+            trace!("have an event!");
+            match event {
+                DatagramEvent::ConnectionEvent(event) => {
+                    self.send_connection_event(handle, event);
+                    continue;
+                }
+                DatagramEvent::NewConnection(connection) => {
+                    debug!("new connection detected!");
+                    break Poll::Ready(Ok((handle, connection)));
+                }
+            }
+        }
+    }
+
+    fn poll_transmit_pending(
+        &mut self,
+        socket: &socket::Socket,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Error>> {
+        let Self { inner, pending, .. } = self;
+        pending
+            .send_packet(cx, socket, &mut || inner.poll_transmit())
+            .map_err(Error::IO)
+    }
+}
+
+type ListenerResult = ListenerEvent<Upgrade, Error>;
+
+#[derive(Debug)]
+pub(super) struct EndpointData {
+    /// The single UDP socket used for I/O
+    socket: Arc<socket::Socket>,
+    /// A `Mutex` protecting the QUIC state machine.
+    inner: Mutex<EndpointInner>,
+    /// The channel on which new connections are sent.  This is bounded in practice by the accept
+    /// backlog.
+    new_connections: mpsc::UnboundedSender<ListenerResult>,
+    /// The channel used to receive new connections.
+    receive_connections: Mutex<Option<mpsc::UnboundedReceiver<ListenerResult>>>,
+    /// The `Multiaddr`
+    address: Multiaddr,
+    /// The configuration
+    config: Config,
+}
+
+type Sender = mpsc::Sender<EndpointMessage>;
+
+#[derive(Debug)]
+pub(crate) struct ConnectionEndpoint {
+    pending: socket::Pending,
+    connection: Connection,
+    handle: ConnectionHandle,
+    channel: Channel,
+    waker: Option<std::task::Waker>,
+}
+
+impl ConnectionEndpoint {
+    /// Notify the endpoint that the handshake has completed,
+    /// and get the certificates from it.
+    ///
+    /// If this returns [`Poll::Pending`], the current task can be worken up by
+    /// a call to [`ConnectionEndpoint::wake`].
+    ///
+    /// Calling this after it has returned `Ready` is erroneous.
+    pub(crate) fn notify_handshake_complete(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<libp2p_core::PeerId, Error>> {
+        if self.connection.is_handshaking() {
+            self.set_waker(cx);
+            return Poll::Pending;
+        }
+        if self.connection.side().is_server() {
+            ready!(self.channel.poll_ready(cx))?;
+            self.channel
+                .start_send(EndpointMessage::ConnectionAccepted)?
+        }
+        let certificate = self
+            .connection
+            .crypto_session()
+            .get_peer_certificates()
+            .expect("we always require the peer to present a certificate; qed");
```

Comment:
This sporadically panics for me with the _distributed-key-value-store_ example and the following patch (I still need to dig deeper):


```diff
diff --git a/examples/distributed-key-value-store.rs b/examples/distributed-key-value-store.rs
index 5bf21855..ea5a5335 100644
--- a/examples/distributed-key-value-store.rs
+++ b/examples/distributed-key-value-store.rs
@@ -33,6 +33,10 @@ use async_std::{io, task};
 use futures::prelude::*;
 use libp2p::kad::record::store::MemoryStore;
 use libp2p::kad::{record::Key, Kademlia, KademliaEvent, PutRecordOk, Quorum, Record};
+use libp2p::quic::{
+    Config,
+    Endpoint,
+};
 use libp2p::{
     NetworkBehaviour,
     PeerId,
@@ -51,8 +55,13 @@ fn main() -> Result<(), Box<dyn Error>> {
     let local_key = identity::Keypair::generate_ed25519();
     let local_peer_id = PeerId::from(local_key.public());
 
-    // Set up a an encrypted DNS-enabled TCP Transport over the Mplex protocol.
-    let transport = build_development_transport(local_key)?;
+    let quic_config = Config::new(&local_key);
+
+    let (quic_endpoint, join_handle) = Endpoint::new(
+        quic_config,
+        "/ip4/0.0.0.0/udp/0/quic".parse().expect("bad address?"),
+    )
+    .expect("I/O error");
 
     // We create a custom network behaviour that combines Kademlia and mDNS.
     #[derive(NetworkBehaviour)]
@@ -66,6 +75,7 @@ fn main() -> Result<(), Box<dyn Error>> {
         fn inject_event(&mut self, event: MdnsEvent) {
             if let MdnsEvent::Discovered(list) = event {
                 for (peer_id, multiaddr) in list {
+                    println!("mdns discovered multiaddr: {:?}", multiaddr);
                     self.kademlia.add_address(&peer_id, multiaddr);
                 }
             }
@@ -109,14 +119,14 @@ fn main() -> Result<(), Box<dyn Error>> {
         let kademlia = Kademlia::new(local_peer_id.clone(), store);
         let mdns = Mdns::new()?;
         let behaviour = MyBehaviour { kademlia, mdns };
-        Swarm::new(transport, behaviour, local_peer_id)
+        Swarm::new(quic_endpoint, behaviour, local_peer_id)
     };
 
     // Read full lines from stdin
     let mut stdin = io::BufReader::new(io::stdin()).lines();
 
     // Listen on all interfaces and whatever port the OS assigns.
-    Swarm::listen_on(&mut swarm, "/ip4/0.0.0.0/tcp/0".parse()?)?;
+    Swarm::listen_on(&mut swarm, "/ip4/0.0.0.0/udp/0/quic".parse()?)?;
 
     // Kick it off.
     let mut listening = false;
```

Reply:
This should be fixed.